code/modules/reagents/chemistry/holder.dm

```diff
@@ -368,22 +282,75 @@
 	handle_reactions()
 	return total_removed //this should be amount unless the loop is prematurely broken, in which case it'll be lower. It shouldn't ever go OVER amount.
 
-/// Removes all reagents from this holder
+/**
+ * Removes all reagents by an amount equal to
+ * [amount specified] / total volume present in this holder
+ * Arguments
+ *
+ * * amount - the volume of each reagent
+ */
+
 /datum/reagents/proc/remove_all(amount = 1)
+	if(!total_volume)
+		return FALSE
+
+	if(!IS_FINITE(amount))
+		stack_trace("non finite amount passed to remove all reagents [amount]")
+		return FALSE
+
+	// Prevents small amount problems, as well as zero and below zero amounts.
+	amount = FLOOR(amount, CHEMICAL_QUANTISATION_LEVEL)
+	if(amount <= 0)
+		return FALSE
+
 	var/list/cached_reagents = reagent_list
-	if(total_volume > 0)
-		var/part = amount / total_volume
+	var/removed_amount = 0
+	var/equal_contribution = amount * (1 / length(cached_reagents))
+	var/final_contribution = 0
+	/**
+	 * when i = 1(1st iteration) each reagent contributes equally to the requested amount
+	 * when i = 2(2nd iteration) each reagent contributes maximum to how much is left over
+	 */
+	for(var/i in 1 to 2)
 		for(var/datum/reagent/reagent as anything in cached_reagents)
-			remove_reagent(reagent.type, reagent.volume * part)
+			var/remove_amount = FLOOR(min(reagent.volume, i == 1 ? equal_contribution : final_contribution), CHEMICAL_QUANTISATION_LEVEL)
+			remove_reagent(reagent.type, remove_amount)
+			removed_amount += remove_amount
+			if(removed_amount >= amount)
+				break
+			if(i == 2)
+				final_contribution = amount - removed_amount
 
-		//finish_reacting() //A just in case - update total is in here - should be unneeded, make sure to test this
-		handle_reactions()
-		return amount
+		final_contribution = amount - removed_amount
+		if(!total_volume || final_contribution < CHEMICAL_QUANTISATION_LEVEL)
+			break
+
+	//finish_reacting() //A just in case - update total is in here - should be unneeded, make sure to test this
```

Comment:
Nope it has stayed commented since the dawn of time itself. I thought i would keep it if someone was interested in testing it in the future but looks like no one will so I'll remove it